[docs] add webhook documentation (#26)
starlinglab/integrity-v2: docs/webhook.md

# Webhook

The webhook service allows users to post files and corresponding attributes to an HTTP endpoint. It supports both CBOR and JSON body formats.

Uploaded files are saved to a local directory with the filename set as the calculated CID. Attributes are posted to the authenticated attribute service. Predefined private attributes are automatically encrypted.

## Setup

Make sure the following config values are set: `AA.*`, `webhook.Host`, `Dirs.Files`, and `Dirs.EncKeys`. `Dirs.Files` is the local directory where the uploaded files are saved as `${CID}`. `Dirs.EncKeys` is where attribute encryption keys are generated/read, with filenames saved in the format of `${CID}_${ATTRIBUTE_KEY}.key`.

An environment variable `JWT_SECRET` should be set as a 32-character secret, which will be used for signing `HS256` authentication JWTs.

## Authenticating with the Webhook

For webhook callers, ensure the config values `webhook.Host` and `webhook.Jwt` are set. `webhook.Jwt` should be a pre-shared `HS256` JWT signed by the webhook host (`JWT_SECRET`).

The JWT token should be set in the `Authorization` HTTP header with a value of `Bearer: ${token}`.

## Endpoints

> ### /generic

Reviewer: ```suggestion
### POST `/generic`
```

#### Body

> - **Type:** `application/x-www-form-urlencoded`

Reviewer: `multipart/form-data` is what is used in the code.

> - **Form Fields:**

Reviewer: ```suggestion
- **Parts:**
```

  | Key      | Description                                                                |
  |----------|----------------------------------------------------------------------------|
  | metadata | File metadata, attributes in key/value pairs, accepts `application/json` or `application/cbor` |

>   | file     | File to be uploaded, `application/octet-stream`                            |

Reviewer: ```suggestion
  | file     | File to be uploaded, must have file name, media type is ignored                            |
```

#### Description

Generic endpoint for uploading and registering a file with attributes.

## Encryption of Private Attributes

> When the attribute key matches the list of predefined private attributes (currently hardcoded values are "private" and "proofmode"), the webhook automatically encrypts the attribute key before registering it on the authenticated attribute service. A 32-byte private key is automatically read or generated from `Dirs.EncKeys`, in the format of `${CID}_${ATTRIBUTE_KEY}.key`.

Reviewer: This description could be expanded to explain that `private` is itself a map of key-value pairs, that get promoted to top-level but encrypted. So that it's obvious that `private` doesn't get stored as one big encrypted attribute.

It makes me wonder if `proofmode` should even be a special key, and if the preprocessor should just put `proofmode` under `private` when sending to the webhook. Thoughts?

Author: On one hand there is no way to know systematically if a field is encrypted now, moving all to `private` would solve this issue
but on the other hand, putting `proofmode` into `private` would mean user need a prior knowledge of `proofmode` is encrypted (compared to say `wacz`), which making parsing hard for new comers
I think it's easier to keep `proofmode` as it but solve the first issue somehow

Reviewer: I see. I guess this way is safer. But I could imagine a deployment where they don't want to encrypt proofmode data at all. It's probably better to just make sure our client puts `proofmode` under `private`, since we control both ends. And other users can do what they want.

I am not sure what you mean by "parsing", do you mean encoding/organization?
